ocaml/typing/mode.ml

```diff
@@ -1220,29 +1312,67 @@ module Comonadic_with_locality = struct
 end
 
 module Monadic = struct
-  let uniqueness m = m
+  module Const = struct
+    include C.Monadic
+  end
 
-  (* secretly just uniqueness *)
-  include Uniqueness
+  module Obj = struct
+    type const = Const.t
+
+    (* Negative solver on the opposite of monadic should give the monadic
+       fragment with original ordering *)
+    module Solver = S.Negative
+
+    let obj = C.Monadic_op
+  end
+
+  include Common (Obj)
 
   type error = [`Uniqueness of Uniqueness.error]
 
   type equate_error = equate_step * error
 
-  let max_with_uniqueness m = S.Negative.disallow_left m
+  let uniqueness m =
+    S.Negative.via_monotone Uniqueness.Obj.obj (Proj (Obj.obj, Uniqueness)) m
 
-  let min_with_uniqueness m = S.Negative.disallow_right m
+  let max_with_uniqueness m =
+    S.Negative.via_monotone Obj.obj (Min_with Uniqueness)
+      (S.Negative.disallow_left m)
 
-  let set_uniqueness_max _ =
-    Uniqueness.max |> S.Negative.disallow_left |> S.Negative.allow_right
+  let min_with_uniqueness m =
+    S.Negative.via_monotone Obj.obj (Max_with Uniqueness)
+      (S.Negative.disallow_right m)
 
-  let set_uniqueness_min _ =
-    Uniqueness.min |> S.Negative.disallow_right |> S.Negative.allow_left
+  let set_uniqueness_max m =
+    S.Negative.via_monotone Obj.obj
+      (C.lift SUniqueness (Const_min Uniqueness_op))
+      (S.Negative.disallow_left m)
 
+  let set_uniqueness_min m =
+    S.Negative.via_monotone Obj.obj
+      (C.lift SUniqueness (Const_max Uniqueness_op))
+      (S.Negative.disallow_right m)
+
+  let zap_to_legacy = zap_to_ceil
```

Comment:
Might be nice to write the explicit tuple here; as it is, this might be easy to forget to update. The legacy mode for contention is *uncontended*.